src/osp/id_registry.h

```diff
@@ -0,0 +1,349 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "hierarchical_bitset.h"
+
+#include <cassert>
+#include <limits>
+#include <utility>
+#include <vector>
+
+namespace osp
+{
+
+template<typename TYPE_T>
+struct type_identity
+{
+    using type = TYPE_T;
+};
+
+template<typename TYPE_T>
+using type_identity_t = typename type_identity<TYPE_T>::type;
+
+template<typename TYPE_T, typename = void>
+struct underlying_int_type;
+
+template<typename TYPE_T>
+struct underlying_int_type< TYPE_T, std::enable_if_t< std::is_enum_v<TYPE_T> > >
```

Comment:
Are we really committing to this formatting style with the spaces inside template parameters? Frankly I'm not a fan of it at all but if we're going to do that we should probably do it everywhere.

Reply:
I go with whatever is best for readability. 

Getting rid of the contents for a moment, it's easy to see the block structure of this line:

`AAA< BBB, CCC< DDD<EEE> > >`

With no spaces at the angles, the block structure is less obvious. Note the space after the comma which makes it seem like a more significant separator compared to the angles:

`AAA<BBB, CCC<DDD<EEE>>>`

The rule I kind of go with, is that if there's a space on one side, then there should be a space on the other side too.

Reply:
It probably comes down to personal preference because personally, seeing the spaces triggers my mental tokenizer to think I'm reading arithmetic comparisons and I find things like `> > >` to be weird and noisy, and not any easier to parse than a densely-packed template argument. Maybe it's just the unfamiliarity. I do agree that sometimes big messy templates are terrible to read, but in this case since the nesting only deepens as you move from left to right it's not very bad. In the cases where it really is that difficult to read, I prefer splitting into multiple lines like

```cpp
struct underlying_int_type<
    TYPE_T,
    std::enable_if_t<std::is_enum_v<TYPE_T>>>
```

although in this case that's sort of overkill.

Anyways, what should we do about it, though? I don't think there's a correct answer here so either we tolerate both ways or commit to using only one. I'm not sure I'll be able to remember to write it this way in the "heat of the moment" although it can always be changed in the code reviews

Reply:
Personally, I don't care enough to enforce a specific style. If it's not horrible, then I'll let it pass. There are some cases where spaces improve readability, and there are some cases where they don't.

A mix of both maybe; ie. the outer-most block having spaces, but the inner ones don't:

`AAA< BBB, CCC<DDD<EEE>>, FFF<GGG> >`